Configure JSON content type for generic webhook RESTClient.

benluddy opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

/kind bug

#### What this PR does / why we need it:

- restclients used to call webhooks are configured with a LegacyCodec (always encodes to json) here: 
https://github.com/kubernetes/kubernetes/blob/f2bed63ca55d24b48fbc3446a5d6e043aa35f138/staging/src/k8s.io/apiserver/pkg/util/webhook/webhook.go#L86
- when encoding a runtime.object to a request body, the encoder is selected here https://github.com/kubernetes/kubernetes/blob/f2bed63ca55d24b48fbc3446a5d6e043aa35f138/staging/src/k8s.io/client-go/rest/request.go#L525. webhook clients select the legacycodec due to a special case for SerializerInfo with empty MediaType: https://github.com/kubernetes/kubernetes/blob/f2bed63ca55d24b48fbc3446a5d6e043aa35f138/staging/src/k8s.io/apimachinery/pkg/runtime/codec.go#L272
- there's no indication to the callsite in `request.go` that encoder selection ignored the requested media type argument, and the Content-Type request header is set to the requested media type argument: https://github.com/kubernetes/kubernetes/blob/f2bed63ca55d24b48fbc3446a5d6e043aa35f138/staging/src/k8s.io/client-go/rest/request.go#L537
- if the rest config is set up to use a content type other than "application/json", webhook request bodies are encoded to JSON but the media type in the Content-Type request header will not match. a webhook server that understands Content-Type and does negotiation will not be able to decode the request body

#### Which issue(s) this PR is related to:
<!--
Please link relevant issues to help with tracking.

To automatically close the linked issue(s) when this PR is merged,
add the word "Fixes" before the issue number or link.
Do not use "Fixes" if the PR is of kind `failing-test` or `flake`.

Reference KEPs when applicable in addition to specific issues.

Examples:
Fixes #<issue number>
<issue link> (issue in a different repository)
KEP: https://github.com/kubernetes/enhancements/issues/<kep-issue-number>

If there is no associated issue, then write "N/A".
-->

#### Special notes for your reviewer:

AFAICT all REST configs used to construct these clients are loaded from kubeconfig files, so the only way to reproduce it is by enabling a client-go feature gate that affects the default when no content type is configured (at least for in-tree uses of `k8s.io/apiserver/pkg/util/webhook`.

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Fixed a bug that caused apiservers to send an inappropriate Content-Type request header to authorization, token authentication, imagepolicy admission, and audit webhooks when the alpha client-go feature gate "ClientsPreferCBOR" is enabled.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
```

benluddy committed:
Configure JSON content type for generic webhook RESTClient.

Authorization, token authentication, imagepolicy admission, and audit webhooks configure RESTClients
that encode to JSON regardless of the ContentType of the provided rest.Config. Because this is
opaque to the RESTClient, configuring a ContentType other than "application/json" results in
requests with JSON-encoded bodies and a non-JSON media type in the Content-Type header. Webhook
servers that respect the Content-Type request header will be unable to decode an object from the
request body.

Explicitly overriding the ContentType of the provided rest.Config fixes this issue and is consistent
with how clients are constructed for conversion and admission webhooks.

benluddy commented:
/kind bug

deads2k commented on staging/src/k8s.io/apiserver/pkg/util/webhook/webhook.go:
```diff
@@ -83,6 +83,7 @@ func NewGenericWebhook(scheme *runtime.Scheme, codecFactory serializer.CodecFact
 	clientConfig := rest.CopyConfig(config)
 
 	codec := codecFactory.LegacyCodec(groupVersions...)
+	clientConfig.ContentType = runtime.ContentTypeJSON
```
Is there any test that ensures this to authn/authz hooks?  If we have such a thing, great.  If we don't, we need one.  Kubelets use this API often.

benluddy replied:
> ensures this

Do you mean a test of the request encoding used for requests made by authn and authz hooks? I will check/add.

benluddy committed:
Test that auth{z,n} hook clients honor Kubelet's request encoding.

Kubelets can make a high volume of SubjectAccessReview and TokenReview requests via authn and authz
webhooks. Its default encoding, Protobuf, mitigates the serialization cost of these requests when
compared to client-go's default encoding, JSON. This test provides some insurance against changes to
webhook client construction that might cause the webhook clients to fail to respect Kubelet's
configured content type.

benluddy commented:
/retest

deads2k commented:
Thanks, that test is very convincing.

/lgtm
/approve

benluddy commented:
/retest-required
